File: 2381-shifting-letters-ii/2381-shifting-letters-ii.cpp
class Solution {
public:
    string shiftingLetters(string s, vector<vector<int>>& shifts) {
        int N = s.length() + 1;
        vector <int> accu(N);
        for(auto &shift: shifts){
            if(shift[2]){
                accu[shift[0]] += 1;
                accu[shift[1] + 1] -= 1;
            }
            else{
                accu[shift[0]] -= 1;
                accu[shift[1] + 1] += 1;
            }
        }
        for(int i = 1; i < N; ++i){
            accu[i] += accu[i - 1];
        }
        string chars = "abcdefghijklmnopqrstuvwxyz";
        for(int i = 0; i < s.length(); ++i){
            int idx = (int)s[i] - 97 + accu[i];
            idx = (idx%26 + 26)%26;
            s[i] = chars[idx];
        }
        return s;
    }
};

File: 2381-shifting-letters-ii/2381-shifting-letters-ii.py
class Solution:
    def shiftingLetters(self, s: str, shifts: List[List[int]]) -> str:
        chars = "abcdefghijklmnopqrstuvwxyz"
        N = len(s) + 1
        accu = [0]*N
        for st, end, move in shifts:
            if move:
                accu[st] += 1
                accu[end + 1] -= 1
            else:
                accu[st] -= 1
                accu[end + 1] += 1
        for i in range(1, N):
            accu[i] += accu[i - 1]
        ans = []
        for i, ch in enumerate(s):
            idx = (ord(ch) - 97 + accu[i])%26
            ans.append(chars[idx])
        return "".join(ans)